metricbeat/module/linux/pageinfo/reader.go

```diff
@@ -0,0 +1,133 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package pageinfo
+
+import (
+	"bufio"
+	"io"
+	"os"
+	"regexp"
+	"strconv"
+
+	"github.com/pkg/errors"
+)
+
+// zones represents raw pagetypeinfo data
+type zones struct {
+	DMA          map[string]map[int]int64
+	DMA32        map[string]map[int]int64
+	Normal       map[string]map[int]int64
+	OrderSummary map[int]int64 `json:"order_summary"`
+}
+
+// buddyInfo emulates /proc/buddyinfo by summing migrate types across orders
+type buddyInfo struct {
+	DMA    map[int]int64
+	DMA32  map[int]int64
+	Normal map[int]int64
+}
+
+// pageInfo represents all the data we get from /proc/pagetypeinfo
+type pageInfo struct {
+	BuddyInfo buddyInfo
+	zones     map[int64]zones
+}
+
+var pageinfoLine = regexp.MustCompile(`Node\s*(\d), zone\s*([a-zA-z0-9]*), type\s*([a-zA-z0-9]*)\s*(\d*)\s*(\d*)\s*(\d*)\s*(\d*)\s*(\d*)\s*(\d*)\s*(\d*)\s*(\d*)\s*(\d*)\s*(\d*)\s*(\d*)`)
+
+// readPageFile reads a PageTypeInfo file and returns the parsed data
+// This returns a massive representation of all the meaningful data in /proc/pagetypeinfo
+func readPageFile(path string) (pageInfo, error) {
```

Comment:
consider to accept `io.Reader` only, so to separate creating/opening the resource lifetime and actual parsing.